marin/rl/envs/math_env.py

```diff
@@ -0,0 +1,224 @@
+
+        # Prepare OpenAI client for the target inference server.
+        self._client = openai.Client(api_key=api_key or openai.api_key, base_url=inference.address)
+
+    # ------------------------------------------------------------------
+    # Main loop
+    # ------------------------------------------------------------------
+
+    async def run(self) -> None:
+        iteration = 0
+        while not await self._should_stop():
+            if self._max_iters is not None and iteration >= self._max_iters:
+                break
+
+            # ------------------------------------------------------------------
+            # Sample a problem (reuse the dataset when exhausted).
+            # ------------------------------------------------------------------
+            if self._example_idx >= len(self._examples):
+                self._rng.shuffle(self._examples)
+                self._example_idx = 0
+            example = self._examples[self._example_idx]
+            self._example_idx += 1
+
+            user_prompt: str = example["prompt"]
+            gt_answer: str = example["answer"]
+
+            # ------------------------------------------------------------------
+            # Call inference server.
+            # ------------------------------------------------------------------
+            completion = self._client.chat.completions.create(
+                model=self._model,
+                messages=[
+                    {"role": "user", "content": user_prompt},
+                ],
+            )
+
+            assistant_msg: str = completion.choices[0].message.content
+
+            # ------------------------------------------------------------------
+            # Reward calculation.
+            # ------------------------------------------------------------------
+            parsed = validate_format(assistant_msg + ">")  # util expects trailing >
+            is_valid = parsed["is_valid"]
+            extracted_answer = parsed["answer"]
+            is_correct = grade_answer(extracted_answer, gt_answer) if is_valid else False
+            reward = float(is_correct)
+
+            # ------------------------------------------------------------------
+            # Build rollout & emit.
+            # ------------------------------------------------------------------
+            turns = [
+                Turn(
+                    message=user_prompt,
+                    role="user",
+                    logprobs=None,
+                    reward=None,
+                    inference_metadata={},
+                ),
+                Turn(
+                    message=assistant_msg,
+                    role="assistant",
+                    logprobs=None,  # logprobs not available via OpenAI client
+                    reward=reward,
+                    inference_metadata={
+                        "model": self._model,
+                        "finish_reason": completion.choices[0].finish_reason,
```

Comment:
Often the env might want to compute multiple rewards (e.g., format rewards, etc.) - does that go here?

Reply:
but does the learner need to know about that?